src/PIL/PpmImagePlugin.py

```diff
@@ -49,71 +53,264 @@ class PpmImageFile(ImageFile.ImageFile):
     format = "PPM"
     format_description = "Pbmplus image"
 
-    def _token(self, s=b""):
+    def _read_magic(self, magic=b""):
         while True:  # read until next whitespace
             c = self.fp.read(1)
-            if not c or c in b_whitespace:
+            if c in B_WHITESPACE:
                 break
-            if c > b"\x79":
-                raise ValueError("Expected ASCII value, found binary")
-            s = s + c
-            if len(s) > 9:
-                raise ValueError("Expected int, got > 9 digits")
-        return s
+            magic += c
+            if len(magic) > 6:  # exceeded max magic number length
+                break
+        return magic
 
-    def _open(self):
+    def _read_token(self, token=b""):
+        def _ignore_comment():  # ignores rest of the line; stops at CR, LF or EOF
+            while self.fp.read(1) not in b"\r\n":
+                pass
+
+        while True:  # read until non-whitespace is found
+            c = self.fp.read(1)
+            if c == b"#":  # found comment, ignore it
+                _ignore_comment()
+                continue
+            if c in B_WHITESPACE:  # found whitespace, ignore it
+                if c == b"":  # reached EOF
+                    raise ValueError("Reached EOF while reading header")
+                continue
+            break
+
+        token += c
+
+        while True:  # read until next whitespace
+            c = self.fp.read(1)
+            if c == b"#":
+                _ignore_comment()
+                continue
+            if c in B_WHITESPACE:  # token ended
+                break
+            token += c
+            if len(token) > 10:
+                raise ValueError(f"Token too long in file header: {token}")
+        return token
 
-        # check magic
-        s = self.fp.read(1)
-        if s != b"P":
-            raise SyntaxError("not a PPM file")
-        magic_number = self._token(s)
-        mode = MODES[magic_number]
+    def _open(self):
+        magic_number = self._read_magic()
+        try:
+            decoder, mode = MODES[magic_number]
+        except KeyError:
+            raise SyntaxError("Not a PPM image file") from None
```

Comment:
Why raise it from `None`?

Reply:
I was considering the underlying cause (`KeyError`) not relevant to the end user.

Reply:
The change this comment was addressing was part of #5121